test/gtest/graphapi_variant_pack.cpp

```diff
@@ -308,86 +284,170 @@ TEST_P(GraphApiVariantPackBuilder, RetrieveAttributes)
     EXPECT_EQ(vPack.getWorkspace(), workspace.value) << "Workspace is set or retrieved incorrectly";
 }
 
+namespace {
+
+using miopen::graphapi::GTestDescriptor;
+using miopen::graphapi::GTestDescriptorAttribute;
+using miopen::graphapi::GTestDescriptorSingleValueAttribute;
+using miopen::graphapi::GTestDescriptorVectorAttribute;
+using miopen::graphapi::GTestGraphApiExecute;
+
+class TensorIds : public GTestDescriptorVectorAttribute<int64_t, char>
+{
+public:
+    TensorIds() = default;
+    TensorIds(const ValidatedVector<int64_t>& vv)
+        : GTestDescriptorVectorAttribute<int64_t, char>(vv.valid,
+                                                        "MIOPEN_ATTR_VARIANT_PACK_UNIQUE_IDS",
+                                                        MIOPEN_ATTR_VARIANT_PACK_UNIQUE_IDS,
+                                                        MIOPEN_TYPE_INT64,
+                                                        MIOPEN_TYPE_CHAR,
+                                                        -1,
+                                                        vv.values)
+    {
+    }
+};
+
+class DataPointers : public GTestDescriptorVectorAttribute<void*, char>
+{
+public:
+    DataPointers() = default;
+    DataPointers(const ValidatedVector<void*>& vv)
+        : GTestDescriptorVectorAttribute<void*, char>(vv.valid,
+                                                      "MIOPEN_ATTR_VARIANT_PACK_DATA_POINTERS",
+                                                      MIOPEN_ATTR_VARIANT_PACK_DATA_POINTERS,
+                                                      MIOPEN_TYPE_VOID_PTR,
+                                                      MIOPEN_TYPE_CHAR,
+                                                      -1,
+                                                      vv.values)
+    {
+    }
+};
+
+class Workspace : public GTestDescriptorSingleValueAttribute<void*, char>
+{
+public:
+    Workspace() = default;
+    Workspace(const ValidatedValue<void*>& vv)
+        : GTestDescriptorSingleValueAttribute<void*, char>(vv.valid,
+                                                           "MIOPEN_ATTR_VARIANT_PACK_WORKSPACE",
+                                                           MIOPEN_ATTR_VARIANT_PACK_WORKSPACE,
+                                                           MIOPEN_TYPE_VOID_PTR,
+                                                           MIOPEN_TYPE_CHAR,
+                                                           2,
+                                                           vv.value)
+    {
+    }
+};
+
+} // namespace
+
+class GraphApiVariantPack : public testing::TestWithParam<GraphApiVariantPackTuple>
+{
+protected:
+    // to be used in the test
+    GTestGraphApiExecute<GTestDescriptorAttribute*> execute;
```

Comment:
question: how is this member `execute` initialized?

Reply:
With "empty" default values as follows:
```C++
template <typename AttributePointer = std::shared_ptr<GTestDescriptorAttribute>>
struct GTestDescriptor
{
    const char* textName               = "";
    miopenBackendDescriptorType_t type = miopenBackendDescriptorType_t(0);
    bool attrsValid                    = false;
    std::vector<AttributePointer> attributes;
};

template <typename AttributePointer = std::shared_ptr<GTestDescriptorAttribute>>
struct GTestGraphApiExecute
{
    GTestDescriptor<AttributePointer> descriptor;
```
SetUp() should set values according to passed parameters.